### ISSUE TITLE
Weird behavior for www.eitf.org

### ISSUE THREAD

**Issue by ThinkChaos** [0xERR0R/blocky#471]
Hi,

I've noticed a weird behavior when querying `www.eitf.org`: blocky returns the `RRSIG`s for the `CNAME` and the domain it points to, even though I'd expect it to handle those internally and just forward the `CNAME` and `A` records.
To get the results I ran the following queries (some output elided for brevity):
```bash
$ dig www.ietf.org @127.0.0.1
[...]
;; QUESTION SECTION:
;; www.ietf.org.       		IN	A

;; ANSWER SECTION:
www.ietf.org.       	1194	IN	CNAME	www.ietf.org.cdn.cloudflare.net.
www.ietf.org.       	1194	IN	RRSIG	CNAME 5 3 1800 20230328000424 20220327230707 40452 ietf.org. MNYE8[...]==
www.ietf.org.cdn.cloudflare.net. 1194	IN	A	104.16.44.99
www.ietf.org.cdn.cloudflare.net. 1194	IN	A	104.16.45.99
www.ietf.org.cdn.cloudflare.net. 1194	IN	RRSIG	A 13 6 300 20220331024317 20220329004317 34505 cloudflare.net. teE5[...]==
[...]
```

This is causing an issue for me because with systemd's resolved pointing to blocky the domain ends up not resolving!
```bash
$ resolvectl query www.ietf.org
www.ietf.org: resolve call failed: Received invalid reply
```

<details>
<summary>Extra logs (that don't add too much information)</summary>

```
$ sudo resolvectl log-level debug
$ journalctl -efu systemd-resolved.service # run the query in another shell
[...]
Cache miss for www.ietf.org IN AAAA
Firing regular transaction 30781 for <www.ietf.org IN AAAA> scope dns on */* (validate=yes).
Using feature level TLS+EDNS0+D0 for transaction 30781.
Using DNS server 127.0.0.1#dns.[redacted] for transaction 30781.
Sending query via TCP since UDP isn't supported or DNS-over-TLS is selected.
Using feature level TLS+EDNS0+D0 for transaction 30781.
Announcing packet size 512 in egress EDNS(0) packet.
Cache miss for www.ietf.org IN A
Firing regular transaction 26063 for <www.ietf.org IN A> scope dns on */* (validate=yes).
Using feature level TLS+EDNS0+D0 for transaction 26063.
Using DNS server 127.0.0.1#dns.[redacted] for transaction 26063.
Sending query via TCP since UDP isn't supported or DNS-over-TLS is selected.
Using feature level TLS+EDNS0+D0 for transaction 26063.
Announcing packet size 512 in egress EDNS(0) packet.
Got message type=method_return sender=org.freedesktop.DBus destination=:1.251 path=n/a interface=n/a member=n/a cookie=12 reply_cookie=15 signature=n/a error
-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.DBus',path='/org/freedesktop/DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0=':1.259' succe
ssfully installed.
Processing incoming packet of size 427 on transaction 30781 (rcode=SUCCESS).
Regular transaction 30781 for <www.ietf.org IN AAAA> on scope dns on */* now complete with <invalid-reply> from network (unsigned; non-confidential).
Processing incoming packet of size 403 on transaction 26063 (rcode=SUCCESS).
Regular transaction 26063 for <www.ietf.org IN A> on scope dns on */* now complete with <invalid-reply> from network (unsigned; non-confidential).
Freeing transaction 30781.
Sent message type=error sender=n/a destination=:1.259 path=n/a interface=n/a member=n/a cookie=17 reply_cookie=2 signature=s error-name=org.freedesktop.resol
ve1.InvalidReply error-message=Received invalid reply
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie
=18 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Freeing transaction 26063.
```
</details>

What I (think I) know and why:
 - not caused by a specific upstream: I have multiple configured and even tried a couple others in case.
 - I tested with both resolved on the same machine as blocky and another. Doesn't impact the results.
 - not linked to the domain being a CNAME: `resolvectl query --cname=false www.ietf.org` has the same result.

I think this is a bug in blocky, or more likely [miekg/dns](https://github.com/miekg/dns), but I wanted to get your opinion before opening a bug there.
I didn't have time to debug blocky and see why the `RRSIG`s are coming through.
Can you reproduce blocky returning the `RRSIG`s with `dig`? Do you think this is indeed a bug in the `dns` module?

Sorry for the long message!

EDIT:

Tried enabling DoH and using curl which seems to give us a better indication of what's wrong:
```bash
$ curl -v --doh-url https://[redacted] www.eitf.org
[...]
* DoH: Too small type A for www.eitf.org
* DoH: Too small type AAAA for www.eitf.org
```

I also set Firefox to use my DoH and while I didn't look for logs, the domain doesn't load like that either.
This all confirms it's definitely not a client issue and more on blocky's side.

---

I also noticed that kdig complains about data missing when using UDP:
```bash
$ kdig www.ietf.org @127.0.0.1
;; WARNING: truncated reply from 127.0.0.1@53(UDP), retrying over TCP
[...]
$ kdig +edns www.ietf.org @127.0.0.1
# No warning
```

I don't think this is the cause of the issue: resolved uses DoT which uses TCP and EDNS, but would rather include it in case I'm wrong!

**Comment by 0xERR0R** [comment]
This is my response (dig with blocky):
```bash
dig www.ietf.org                                                                                                     

; <<>> DiG 9.18.0 <<>> www.ietf.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 1656
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;www.ietf.org.                  IN      A

;; ANSWER SECTION:
www.ietf.org.           1278    IN      CNAME   www.ietf.org.cdn.cloudflare.net.
www.ietf.org.cdn.cloudflare.net. 180 IN A       104.16.44.99
www.ietf.org.cdn.cloudflare.net. 180 IN A       104.16.45.99

;; Query time: 33 msec
;; SERVER: 192.168.178.X#53(192.168.178.X) (UDP)
;; WHEN: Wed Mar 30 21:37:20 CEST 2022
;; MSG SIZE  rcvd: 118
```

Which upstream server(s) are you using?

**Comment by 0xERR0R** [comment]
Sorry for closing, wrong button click ;)

**Comment by ThinkChaos** [comment]
Hmm I must've been tired yesterday night, it looks like it's indeed an issue with a specific upstream: `anycast.uncensoreddns.org`.
I might have time to debug this later, otherwise I'll look into it more tomorrow.

**Comment by ThinkChaos** [comment]
I determined this only happens when I query through resolved. I haven't managed to replicate this with dig.

I also tried testing against a known bad DNSSEC domain and found some more interesting stuff:

```bash
# First query, resolved complains
$ resolvectl query --cache=false sigfail.verteiltesysteme.net
sigfail.verteiltesysteme.net: resolve call failed: Received invalid reply

# Second query, it gets a response
$ resolvectl query --cache=false sigfail.verteiltesysteme.net
sigfail.verteiltesysteme.net: 134.91.78.139    -- link: enp6s18
                              2001:638:501:8efc::139 -- link: enp6s18

-- Information acquired via protocol DNS in 1.9ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: yes
-- Data from: network
```

The above is with `anycast.uncensoreddns.org` upstream from blocky. If I use `dot.sb`, I get the first lookup returns a different error: `resolve call failed: DNSSEC validation failed: failed-auxiliary`, and the second also works.

The fact that 2nd queries return a valid response looks like #455, but I'm not sure.

Tomorrow I'll dump the DNS packets blocky receives to compare dig and resolved.
I tried Wireshark but it wasn't helpful as even when knowing the cert, Diffie-Helman is being used which provides forward secrecy.

Knowing the contents of packets I hope to be able to determine if this is linked to #455: does resolved use `+cd`?
From its logs I don't think so, especially since I configured it with `DNSSEC=true`, but I'd rather be sure.

If I don't spot a difference between dig and resolved, I'll move onto debugging blocky (and the upstream dns module) to figure out what's going on. I should have more time tomorrow, and this week-end to get to the bottom of this.

**Comment by ThinkChaos** [comment]
I dumped the messages and indeed, resolved uses the `cd` flag, which causes cache poisoning as noted in #455.
I'm pretty sure this causes both issues I noticed, the only difference being who poisons the cache first: resolved first leads to dig getting the RRSIG in the response; and dig first causing systemd not to resolve the domain (because the cached response validated dnssec properly).

I'll leave this open for now: I want to find a solution so that resolved and blocky work well together.
All the settings I've tried until now either end in resolved sometimes erroring when it shouldn't, or preventing DNSSEC from working entirely because of the `cd` flag.
